[PATCH V3 00/24] Battery and volume RAM cache

cathay4t opened the pull request:
* New features:
    * Query battery(including super capacitor)
    * Query volume RAM cache settings:
        * Physical disk cache policy.
        * Write cache policy and status.
        * Read cache policy and status.
    * Change volume RAM cache policy.

 * Supported plugins:
    * MegaRAID
    * HPSA
    * Sim and Simc

Please be advised SSD cache will be another PR in version 1.4.

tasleson commented:
Hi Gris,
Please take a look at cache support in LVM to see if what you have here fits into that as well.

cathay4t commented:
@tasleson If convenient, could review the API design in `_client.py` file while I am coding implementation on hpsa and megaraid plugins? It's OK if you prefer to review then after I get PR all ready.

joehandzik commented on plugin/hpsa/hpsa.py:
```diff
@@ -863,3 +880,106 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+           ctrl_data['Battery/Capacitor Status'] == 'OK':
+            flag_battery_ok = True
+
+        # TODO(Gris Ge): support SSD cache(HP SmartCache).
```
I see this comment here...I agree with you that it'd be great to support SmartCache in some way, but I don't necessarily think it should fold in with the BBWC implementation.

joehandzik commented:
Also, FWIW, my PR's unit test run is failing with the same error that yours is...did something break within the last few commits?

joehandzik commented:
Nevermind...I think yours is right to fail. Lemme see what I can figure out on mine.

cathay4t commented:
@tasleson
I got the python and plugin part done.  Could you review the last patch above while I am coding the C part?

About LVM cache, we could add `lsm.Client.volume_ssd_cache_info()` after 1.3 release.

tasleson commented:
@cathay4t Yep, I'll take a look as soon as I figure out why my `plugin_test.py` changes tanked the automated tests in the lab :-/

joehandzik commented on plugin/hpsa/hpsa.py:
```diff
@@ -231,6 +231,17 @@ def _disk_status_of(hp_disk, flag_free):
+        return Disk.LINK_TYPE_ATA
+    elif disk_interface == 'SAS':
+        return Disk.LINK_TYPE_SAS
+
```
Is there not a 'Solid State SAS'?

cathay4t replied:
@joehandzik I don't have SAS SSD connected to our hpsa card and failed to find any document indicate possible values of 'Interface Type'. Above is what I find in our lab.
If that is the property value for SAS SSD, I could add it. Could you confirm that with me or even better provide output of `hpssacli ctrl all show config detail` on controller with SAS SSD?
Thank you.

joehandzik replied:
I'll check it for you, @cathay4t. Didn't have a chance to get my hands on a SAS SSD today. :/

joehandzik replied:
Hey @cathay4t, for reasons explained elsewhere I'm still unable to actually test this on a Solid State SAS drive, but take a look at this link for an example of a Solid State SAS output from hpssacli: http://serverfault.com/questions/586626/hp-ssd-on-proliant-dl360p-gen8-p420i-controller-no-trim

joehandzik commented:
@cathay4t Just a note here that I intend to leverage this PR to cover one of the requirements that cropped up for a use case of ours (the acceleration mode of a volume; controller cache, etc). Should cut down on the extra PRs that I need to spit out in the next week or two. I'll try to get a review in over the next day or two.

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+
+            sys_id = ctrl_data["Serial Number"]
+
+            battery_type = Battery.TYPE_UNKNOWN
```
This assignment isn't needed, the variable `battery_type` either gets set to TYPE_CAPACITOR or TYPE_CHEMICAL or an exception is raised.

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+
+        for ctrl_name in ctrl_all_show.keys():
+            ctrl_data = ctrl_all_show[ctrl_name]
+            if 'Battery/Capacitor Count' not in ctrl_data or \
```
The string `'Battery/Capacitor Count'` gets used a number of times, I would suggest placing it in a constant so when/if it needs to get changed it will be easier.

joehandzik replied:
I'm not opposed to that, but keep in mind it'll likely be a situation more like the bug I fixed the other day for cache modules. If a new string needed to be handled, we'd still need to handle the old string too, this would probably never go away.

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+                    ErrorNumber.PLUGIN_BUG,
+                    "Got unknown battery type: %s for system %s" %
+                    (ctrl_data["Cache Backup Power Source:"], sys_id))
+
```
Do we want to make the plugin un-usable if/when a new battery type is added or do we want to return `TYPE_UNKNOWN` instead?

cathay4t replied:
`TYPE_UNKNOWN` is better. Thanks.

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+        lsm_bats = self.batteries()
+        for lsm_bat in lsm_bats:
+            if lsm_bat.status == Battery.STATUS_OK:
+                flag_battery_ok = True
```
I'm not familiar with the hardware, but if it contains more than one battery is it OK to assume everything is fine for this volume or would it be better to assert that all batteries need to be good before making the conclusion that the battery is all OK?

cathay4t replied:
@joehandzik Could you answer these questions:
 * Can hpsa support multiple batteries or mix type(capacitor and battery) batteries?
 * Will hpsa enable write cache if one of batteries is in good mode while others don't?

Thank you.

joehandzik replied:
Multiple capacitors can definitely be supported within a chassis (it's required for multiple controllers that need a power safe cache prior to Gen9), but you'd only have one per controller/system.

In Gen9, we went to a shared battery model (hedged our bets that we'd need it for nvdimms in Gen9, and we were right). As far as I've heard so far today from asking a few people, there aren't any chassis that support multiple batteries. But if there were in the future, the logic is such that you need a certain amount of "loads" in the battery for whatever hardware that is present in the server. If you don't have enough loads to cover all the hardware in your box that would need the power, you're not power safe. So, while we don't deal with situations where one battery is good and another is bad today, you can envision a world were we would. My assumption is that hpsa would enable write cache as long as there was enough power spread across the batteries in a system to cover a power outage. But, then I'd also expect the battery status to be reported as OK, so what you're doing here would be fine in this case.

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+
+        return [write_cache_policy, write_cache_status, read_cache_policy,
+                read_cache_status, phy_disk_cache]
+
```
If `flag_ram_ok` is false and `if ctrl_data.get('No-Battery Write Cache', '') == 'Enabled'` is false, `write_cache_status` and `read_cache_statue` will be undefined.  `phy_disk_cache` will be undefined if `ctrl_data['Drive Write Cache']` value not in the set `['Disable', 'Enabled']`

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+        for lsm_disk in lsm_disks:
+            if lsm_disk.disk_type == Disk.TYPE_SSD and \
+               lsm_disk.id in disk_ids:
+                return True
```
Wouldn't it be better to iterate through the disks that make up the pool instead of iterating though all the disks and then checking if the disk is in the pool?

cathay4t replied:
Will use method in http://stackoverflow.com/questions/1388818/how-can-i-compare-two-lists-in-python-and-return-matches and depend on python's optimization.

tasleson commented on plugin/megaraid/megaraid.py:
```diff
@@ -229,6 +229,108 @@ def _disk_link_type_of(disk_show_basic_dict):
+    "Degraded (need attention)": Battery.STATUS_DEGRADED,
+    # ^ These values are copy from LSI document, except Optimal all other
+    #   value is not tested
+}
```
All other values are not tested

tasleson commented on plugin/megaraid/megaraid.py:
```diff
@@ -229,6 +229,108 @@ def _disk_link_type_of(disk_show_basic_dict):
+def _mega_bbu_to_lsm(sys_id, bbu_show_all_output):
+    design_info = _fix_bbu_cv_output(bbu_show_all_output['BBU_Design_Info'])
+    bbu_info = _fix_bbu_cv_output(bbu_show_all_output['BBU_Info'])
+    cap_info = _fix_bbu_cv_output(bbu_show_all_output['BBU_Capacity_Info'])
```
`cap_info` is not used?

tasleson commented on plugin/megaraid/megaraid.py:
```diff
@@ -229,6 +229,108 @@ def _disk_link_type_of(disk_show_basic_dict):
+    design_info = _fix_bbu_cv_output(cv_show_all_output['Design_Info'])
+    cv_info = _fix_bbu_cv_output(cv_show_all_output['Cachevault_Info'])
+    cap_info = _fix_bbu_cv_output(cv_show_all_output['GasGaugeStatus'])
+
```
`cap_info` is not used?

tasleson commented on plugin/megaraid/megaraid.py:
```diff
@@ -304,6 +406,13 @@ def capabilities(self, system, flags=Client.FLAG_RSVD):
+        cap.set(Capabilities.VOLUME_WRITE_CACHE_POLICY_SET_WRITE_BACK)
+        cap.set(Capabilities.VOLUME_WRITE_CACHE_POLICY_SET_AUTO)
+        cap.set(Capabilities.VOLUME_WRITE_CACHE_POLICY_SET_WRITE_THROUGH)
+        cap.set(Capabilities.VOLUME_READ_CACHE_POLICY_SET)
```
Why advertise this feature `VOLUME_READ_CACHE_POLICY_SET` when the plugin will return `NO_SUPPORT`?

tasleson commented on plugin/megaraid/megaraid.py:
```diff
@@ -830,3 +940,180 @@ def volume_raid_create(self, name, raid_type, disks, strip_size,
+            raise LsmError(
+                ErrorNumber.INVALID_ARGUMENT,
+                "Illegal input volume argument: missing plugin_data property")
+
```
How about replacing this code with a call to `_vd_path_of_lsm_vol` ?

tasleson commented on plugin/sim/simarray.py:
```diff
@@ -1629,10 +1663,27 @@ def sim_exp_delete(self, sim_exp_id):
-        Return a list of sim_tgt dict.
+        return a list of sim_tgt dict.
+        """
+        return self._get_table('tgts', backstore.tgt_key_list)
```
`Backstore.TGT_KEY_LIST` ?  Not finding `backstore.tgt_key_list`

tasleson commented on plugin/sim/simarray.py:
```diff
@@ -134,6 +134,9 @@ class BackStore(object):
     SYS_MODE = System.MODE_HARDWARE_RAID
+    DEFAULT_WRITE_CACHE_POLICY = Volume.WRITE_CACHE_POLICY_AUTO
+    DEFAULT_READ_CACHE_POLICY = Volume.READ_CACHE_POLICY_ENABLED
+    DEFAULT_PHYSIACL_DISK_CACHE = Volume.PHYSICAL_DISK_CACHE_DISABLED
```
Typo `DEFAULT_PHYSICAL_DISK_CACHE`

tasleson commented on test/plugin_test.py:
```diff
@@ -1473,6 +1475,96 @@ def test_invalid_uri(self):
+                lsm_vols = self.volumes(system_id=s.id)
+                if len(lsm_vols) == 0:
+                    lsm_vol = self._volume_create(s.id)[0]
+                    flag_created = true
```
`True` not `true`

tasleson commented on test/plugin_test.py:
```diff
@@ -1473,6 +1475,96 @@ def test_invalid_uri(self):
+        flag_created = False
+        for s in self.systems:
+            cap = self.c.capabilities(s)
+            if supported(cap, [Cap.VOLUME_CACHE_INFO]):
```
You should check for volume create and volume delete capability too, if you have no volumes

tasleson commented on test/plugin_test.py:
```diff
@@ -1473,6 +1475,96 @@ def test_invalid_uri(self):
+    def test_volume_cache_set(self):
+        flag_tested = False
+        for s in self.systems:
+            cap = self.c.capabilities(s)
```
Check capabilities for all operations you need for test, volume create/delete etc.

tasleson commented on test/plugin_test.py:
```diff
@@ -1473,6 +1475,96 @@ def test_invalid_uri(self):
+                                lsm.Volume.READ_CACHE_POLICY_DISABLED)
+
+            self._volume_delete(lsm_vol)
+            flag_tested = True
```
This flag only really proves if the hardware support VOLUME_CACHE_INFO as all the other tests could fail to even test if not supported.

cathay4t replied:
Will fix that. Thanks.

tasleson commented:
@cathay4t @joehandzik

I see `status` and `policy` information.  However, when I look at them they seem quite similar, eg. 

```python
lsm.Volume.WRITE_CACHE_POLICY_WRITE_BACK
lsm.Volume.WRITE_CACHE_POLICY_AUTO
lsm.Volume.WRITE_CACHE_POLICY_WRITE_THROUGH
lsm.Volume.WRITE_CACHE_POLICY_UNKNOWN
lsm.Volume.WRITE_CACHE_STATUS_WRITE_THROUGH
lsm.Volume.WRITE_CACHE_STATUS_WRITE_BACK
lsm.Volume.WRITE_CACHE_STATUS_UNKNOWN

lsm.Volume.READ_CACHE_POLICY_ENABLED
lsm.Volume.READ_CACHE_POLICY_DISABLED
lsm.Volume.READ_CACHE_POLICY_UNKNOWN
lsm.Volume.READ_CACHE_STATUS_ENABLED
lsm.Volume.READ_CACHE_STATUS_DISABLED
lsm.Volume.READ_CACHE_STATUS_UNKNOWN
```
What exactly are we trying to convey?

cathay4t commented:
@tasleson `policy` is what we could change. `status` is what actual effective.

Example: write cache policy == AUTO, with battery OK, `status` is write back, with batter down, `status` is write through.

Same thing also apply to read cache when cache module installed or not.

Since we are providing functions to change cache policy, I think it could be better to provide property on what we changed.

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1029,6 +1030,102 @@ int LSM_DLL_EXPORT lsm_volume_ident_led_set(lsm_connect *c,
                                               lsm_flag flags);
+/**
+ * Get a list of batteries for this storage system.
+ * The supper capacitor will also be included.
```
*super

Maybe change the whole line to something like: "When present, super capacitors will also be included."

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1029,6 +1030,102 @@ int LSM_DLL_EXPORT lsm_volume_ident_led_set(lsm_connect *c,
+ *                            cache hardware found. In write back mode, I/O
+ *                            write data is done asynchronously to the cache and
+ *                            to the backing store(normally disk), the
+ *                            acknowledgement will send back to I/O initiator
```
Change to "the acknowledgement will be sent back to the I/O initiator once the data has been written to the cache".

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1029,6 +1030,102 @@ int LSM_DLL_EXPORT lsm_volume_ident_led_set(lsm_connect *c,
+ *                          * LSM_VOLUME_READ_CACHE_POLICY_ENABLED
+ *                              Read cache is enabled, when reading I/O on
+ *                              previous unchanged written I/O or read I/O in
+ *                              cache will be return I/O initiator immediately
```
will be returned to the I/O initiator immediately without checking the backing store

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1029,6 +1030,102 @@ int LSM_DLL_EXPORT lsm_volume_ident_led_set(lsm_connect *c,
+ *                          * LSM_VOLUME_PHYSICAL_DISK_CACHE_USE_DISK_SETTING
+ *                              Physical disk cache is determined by the disk
+ *                              vendor vis physical disks' SCSI caching mode
+ *                              page(0x08 page). It's strongly suggest to
```
strongly suggested

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1029,6 +1030,102 @@ int LSM_DLL_EXPORT lsm_volume_ident_led_set(lsm_connect *c,
+ *                              Physical disk cache disabled.
+ *                          * LSM_VOLUME_PHYSICAL_DISK_CACHE_USE_DISK_SETTING
+ *                              Physical disk cache is determined by the disk
+ *                              vendor vis physical disks' SCSI caching mode
```
via instead of vis?

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1004,6 +1004,84 @@ typedef int (*lsm_plug_volume_ident_led_set) (lsm_plugin_ptr c,
+ *                            cache hardware found. In write back mode, I/O
+ *                            write data is done asynchronously to the cache and
+ *                            to the backing store(normally disk), the
+ *                            acknowledgement will send back to I/O initiator
```
Change to "the acknowledgement will be sent back to the I/O initiator once the data has been written to the cache".

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1004,6 +1004,84 @@ typedef int (*lsm_plug_volume_ident_led_set) (lsm_plugin_ptr c,
+ *                          * LSM_VOLUME_READ_CACHE_POLICY_ENABLED
+ *                              Read cache is enabled, when reading I/O on
+ *                              previous unchanged written I/O or read I/O in
+ *                              cache will be return I/O initiator immediately
```
will be returned to the I/O initiator immediately without checking the backing store

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -268,6 +273,22 @@ typedef enum {
+    LSM_DISK_LINK_TYPE_ADT = 7,
+    LSM_DISK_LINK_TYPE_ATA = 8,
+    LSM_DISK_LINK_TYPE_USB = 9,
+    LSM_DISK_LINK_TYPE_SOP = 0xa,
```
Some folks seem to not like your abrupt switch to hex in places like this. I personally don't care, but pointing it out in case you want to swap back to decimal.

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -340,6 +372,56 @@ typedef enum {
 #define LSM_SYSTEM_CACHE_PCT_NO_SUPPORT             -2
 #define LSM_SYSTEM_CACHE_PCT_UNKNOWN                -1
 
+#define LSM_VOLUME_WRITE_CACHE_POLICY_UNKNOWN       0
```
We're using -1 for unknown in some cases, and 0 here. Should we pick a consistent value for *_UNKNOWN macros? At least for new ones, anyway?

cathay4t replied:
I intend to pick `1 (1 << 0)` for UNKNOWN and `2 (1 << 2)` for NO_SUPPORT in order to support bitmap used in status.

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -340,6 +372,56 @@ typedef enum {
+ *   recharging again.
+ */
+#define LSM_BATTERY_STATUS_DEGRADED                 1 << 6
+/* ^ Battery is in degraded mode, need attention.
```
*needs

joehandzik commented on c_binding/lsm_convert.cpp:
```diff
@@ -156,7 +156,7 @@ lsm_disk *value_to_disk(Value & disk)
         if ((rc != NULL) && std_map_has_key(d, "disk_location") &&
             (d["disk_location"].asC_str()[0] != '\0')) {
-            
+
```
Any particular reason for the newline? It's not present in the new attribute checks you just added...

cathay4t replied:
It's my vim macros automatically correcting white spaces.

joehandzik commented on c_binding/lsm_convert.hpp:
```diff
@@ -300,4 +301,29 @@ int LSM_DLL_LOCAL values_to_uint32_array(Value & value,
+
+/**
+ * Converts a vector of battery values to an array.
+ * @param[in] battery_values       Vector of values that represents batterys
```
Should change to "batteries" not "batterys" (see line 326). You'll need to update this throughout the comments if the misspelling exists elsewhere.

joehandzik commented on c_binding/lsm_local_disk.c:
```diff
@@ -29,11 +32,41 @@
+#define _T10_SPC_VPD_SUP_VPD_PGS                         0x00
+/* ^ SPC-5 rev7 7.7.16 Supported VPD Pages VPD page */
+#define _T10_SPC_VPD_SUP_VPD_PGS_MAX_PAGE_LEN            0xff + 1 + 4
+/* ^ There are 256(0xff + 1) page codes, each only take 1 byte, we only
```
*takes

joehandzik commented on c_binding/lsm_local_disk.c:
```diff
@@ -29,11 +32,41 @@
+/* ^ SPC-5 rev7 7.7.16 Supported VPD Pages VPD page */
+#define _T10_SPC_VPD_SUP_VPD_PGS_MAX_PAGE_LEN            0xff + 1 + 4
+/* ^ There are 256(0xff + 1) page codes, each only take 1 byte, we only
+ *   need 256 bytes to store them. Addition 4 bytes for VPD 0x00 page header.
```
*Additional

joehandzik commented on c_binding/lsm_local_disk.c:
```diff
@@ -96,13 +137,22 @@ struct t10_vpd83_dp {
 };
 
+struct t10_sbc_vpd_bdc {
+    uint8_t we_dont_care_0;
```
Maybe change to something like "unused" or "ignored"?

joehandzik commented on plugin/hpsa/hpsa.py:
```diff
@@ -918,3 +955,222 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+            # Check whether we got SSD volume
+            if self._is_ssd_volume(volume):
+                raise LsmError(
+                    ErrorNumber.NO_SUPPORT,
```
The write cache policy's error message is accurate, this error message is not. Also, keep in mind that we could (should?) add a method to disable Smart Path (but call it something generic, like "bypass" or something). Not sure that we'll get it in this release, but add it to a backlog somewhere.

joehandzik commented on python_binding/lsm/_client.py:
```diff
@@ -1427,3 +1427,320 @@ def volume_ident_led_clear(self, volume, flags=FLAG_RSVD):
+        Version:
+            1.3
+        Usage:
+            Query batteries. The supper capacitor will also be included.
```
*super

joehandzik commented on python_binding/lsm/_client.py:
```diff
@@ -1427,3 +1427,320 @@ def volume_ident_led_clear(self, volume, flags=FLAG_RSVD):
+                        hardware found.
+                        In write back mode, I/O write data is done
+                        asynchronously to the cache and to the backing
+                        store(normally disk), the acknowledgement will send back
```
Same comments as the previous two occurrences

joehandzik commented on python_binding/lsm/_client.py:
```diff
@@ -1427,3 +1427,320 @@ def volume_ident_led_clear(self, volume, flags=FLAG_RSVD):
+                    * lsm.Volume.READ_CACHE_POLICY_ENABLED
+                        Read cache is enabled, when reading I/O on previous
+                        unchanged written I/O or read I/O in cache will be
+                        return I/O initiator immediately without checking
```
Same comments as the previous two occurrences

joehandzik commented on python_binding/lsm/_client.py:
```diff
@@ -1427,3 +1427,320 @@ def volume_ident_led_clear(self, volume, flags=FLAG_RSVD):
+        Version:
+            1.3
+        Usage:
+            Change the setting of RAM physical disk cache of certain volume.
```
+            Change the setting of RAM physical disk cache of the specified volume.
 +            On some products (like HPE SmartArray), this action will be effective
 +            at the system level which means that even though you are requesting a change
 +            on a certain volume, the change will apply to all other volumes on the same
 +            controller(system).

joehandzik commented on python_binding/lsm/_client.py:
```diff
@@ -1427,3 +1427,320 @@ def volume_ident_led_clear(self, volume, flags=FLAG_RSVD):
+            lsm.Capabilities.VOLUME_PHYSICAL_DISK_CACHE_SET
+                Allow changing physical disk cache.
+            lsm.Capabilities.VOLUME_PHYSICAL_DISK_CACHE_SET_SYSTEM_LEVEL
+                Indicate this action will change system settings which
```
+                Indicates that this action will change system settings which
 +                are effective for all volumes in this system.
 +                For example, on HPE SmartArray, the physical disk cache
 +                setting is a controller level setting.

joehandzik commented on python_binding/lsm/_client.py:
```diff
@@ -1427,3 +1427,320 @@ def volume_ident_led_clear(self, volume, flags=FLAG_RSVD):
+        Usage:
+            Change the RAM write cache policy on certain volume.
+            If lsm.Capabilities.VOLUME_WRITE_CACHE_POLICY_SET_IMPACT_READ
+            is supported(e.g. HP SmartArray), the changes on write cache policy
```
Just change all the HP mentions to HPE in this pull if you could. I'll submit a pull when I can to update mentions as unintrusively as possible. :)

joehandzik commented on tools/lsmcli/cmdline.py:
```diff
@@ -221,14 +226,29 @@ def _get_item(l, the_id, friendly_name='item', raise_error=True):
+write_cache_policy_types = ['WB', 'AUTO', 'WT']
+write_cache_policy_help = 'Write cache policy: ' + \
+                          ', '.join(write_cache_policy_types) + \
+                          'which stand by "write back", "auto", "write through"'
```
*stand for

BlaineEXE commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1044,6 +1045,172 @@ int LSM_DLL_EXPORT lsm_system_read_cache_pct_update(lsm_connect *c,
 
+/**
+ * Get a list of batteries for this storage system.
+ * The supper capacitor will also be included.
```
'supper' should be changed to 'super'

BlaineEXE commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1044,6 +1045,172 @@ int LSM_DLL_EXPORT lsm_system_read_cache_pct_update(lsm_connect *c,
+ * @param [out]     batteries       An array of lsm_battery types
+ * @param [out]     count           Number of batteries
+ * @param [in]      flags           Reserved set to zero
+ * @return LSM_ERR_OK on success else error reason
```
I'm confused by the phrase 'error reason'. It returns an int type. I'm assuming this means it returns an enumerated error type matching LSM_ERR_<reason> by the success value.

tasleson replied:
@BlaineEXE  The function returns an integer.  If the value == LSM_ERR_OK the operation was successful.  Any other value is the error reason as specified in the header file `libstoragemgmt_error.h`.  You can also retrieve additional information by calling `lsm_error_last_get`.

BlaineEXE replied:
I think the function documentation could benefit from stating that information explicitly rather than leaving the user confused (as I was) and making them find the correct information in the codebase.

joehandzik replied:
@BlaineEXE (aka Lego Spiderman): I would encourage you to create a sample pull request that clarifies the point adequately for you. If we can all agree on an update, then we can cascade your change throughout the codebase. Sound fair?

BlaineEXE replied:
Sounds good. I'm working on that with my enclosure support. Will that be sufficient, or should I separate it entirely?

tasleson replied:
@BlaineEXE Please make it separate.

cathay4t commented:
@joehandzik @tasleson PR is ready for review.
I manually tested on MegaRAID and HPSA for both SSD and HDD disks.

Thank you as always.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1044,6 +1045,23 @@ int LSM_DLL_EXPORT lsm_system_read_cache_pct_update(lsm_connect *c,
+int LSM_DLL_EXPORT lsm_battery_list(lsm_connect *conn,
+                                    const char *search_key,
+                                    const char *search_value,
+                                    lsm_battery ***bs, uint32_t *count,
```
Other functions in the library use the syntax `<type> **name[]` when returning an array of objects, we should stay consistent.

cathay4t replied:
Sure. I will change that.
Just wondering why we are using that when `lsm_battery *bs[] = NULL` will cause compile error.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_battery.h:
```diff
@@ -0,0 +1,108 @@
+ * @param count         Count of battery array.
+ * @return LSM_ERR_OK on success, else error reason.
+ */
+int LSM_DLL_EXPORT lsm_battery_record_array_free(lsm_battery **bs,
```
`lsm_battery *bs[]`

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1030,6 +1031,71 @@ struct lsm_ops_v1_2 {
+typedef int (*lsm_plug_battery_list) (lsm_plugin_ptr c,
+                                      const char *search_key,
+                                      const char *search_val,
+                                      lsm_battery ***bs,
```
`lsm_battery **bs[]`

tasleson commented on c_binding/lsm_plugin_ipc.cpp:
```diff
@@ -2933,3 +2938,72 @@ void lsm_plug_target_port_search_filter(const char *search_key,
+static void get_batteries(int rc, lsm_battery **bs, uint32_t count,
+                          Value &response)
+{
+    uint32_t i = 0;
```
My preference would be to declare `i` in the for loop to minimize scope.

cathay4t replied:
I intend to declare and initialize variables at the beginning of each function.

Using in-loop declaration requires `-std=c99`, or you will get error on RHEL 6 which use C90 by default:
```
error: ‘for’ loop initial declarations are only allowed in C99 mode
```

tasleson replied:
On 04/28/2016 02:06 AM, Gris Ge wrote:
>> @@ -2933,3 +2938,72 @@ void lsm_plug_target_port_search_filter(const char *search_key,
>>          }
>>      }
>>  }
>> +
>> +static void get_batteries(int rc, lsm_battery **bs, uint32_t count,
>> +                          Value &response)
>> +{
>> +    uint32_t i = 0;
> 
> I intend to declare and initialize variables at the beginning of each function.
> 
> Using in-loop declaration requires `-std=c99`, or you will get error on RHEL 6 which use C90 by default:
> ```
> error: ‘for’ loop initial declarations are only allowed in C99 mode
> ```

Agreed, but this is a c++ file being compiled with a c++ compiler which
does support this.

-Tony

> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/libstorage/libstoragemgmt/pull/16/files/b9212a274cf79babb271268fa6f9dee41c2769b3..4de1a915fdda3c89bf07968084e8d7fe7fdf89a3#r61381537
>

tasleson commented on c_binding/lsm_plugin_ipc.cpp:
```diff
@@ -2933,3 +2938,72 @@ void lsm_plug_target_port_search_filter(const char *search_key,
+
+void lsm_plug_battery_search_filter(const char *search_key,
+                                    const char *search_value,
+                                    lsm_battery **bs, uint32_t *count)
```
`lsm_battery *bs[]`

tasleson commented on c_binding/lsm_mgmt.cpp:
```diff
@@ -2610,3 +2613,43 @@ int lsm_system_read_cache_pct_update(lsm_connect * c, lsm_system * system,
+                             lsm_battery **bs[], uint32_t *count)
+{
+    if (LSM_ERR_OK == rc && Value::array_t == response.valueType()) {
+        rc = value_array_to_batteries(response, bs, count);
```
We need to wrap the call to `value_array_to_batteries` with a `try.. catch` see  `get_volume_array`.  This is because `value_to_battery` can raise an exception that must be caught.

tasleson commented on c_binding/lsm_datatypes.cpp:
```diff
@@ -2208,6 +2209,96 @@ char *wwpn_convert(const char *wwpn)
+CREATE_PRO_GETTER_FUNC(lsm_battery_type_get, lsm_battery *,
+                       LSM_IS_BATTERY, lsm_battery_type, type,
+                       LSM_BATTERY_TYPE_UNKNOWN);
+
```
We have a `#define` already `MEMBER_FUNC_GET`, can we re-use that one instead of creating the new one `CREATE_PRO_GETTER_FUNC`?

cathay4t replied:
Sorry, I didn't noticed  that. Will use it.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -367,6 +372,35 @@ typedef enum {
+#define LSM_BATTERY_STATUS_DEGRADED                 1 << 6
+/** Battery is having hardware error or end of life. */
+#define LSM_BATTERY_STATUS_ERROR                    1 << 7
+
```
All the other `#define` for bitfield values are declared using `0x0000000000000001`, `0x0000000000000002`, please make consistent.

cathay4t replied:
Sure thing.

tasleson commented on plugin/hpsa/hpsa.py:
```diff
@@ -1009,3 +1016,37 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
+                        battery_type, battery_status, sys_id,
+                        _plugin_data=None))
+
+        return lsm_bs
```
shouldn't this be:

```python
return search_property(lsm_bs, search_key, search_value)
```

tasleson commented on c_binding/lsm_mgmt.cpp:
```diff
@@ -2653,3 +2653,66 @@ int lsm_battery_list(lsm_connect *c, const char *search_key,
+
+        rc = LSM_ERR_INVALID_ARGUMENT;
+        goto out;
+    }
```
I know people like having a single exit point, but this early in the validation process I'm thinking it would be easier and cleaner to just return here and above with the invalid argument.  That would alleviate the need for having all the if checks below when clearing all the `"out"` variables.

cathay4t replied:
I don't quite understand your suggestion. You mean remove `out` and leave them as `dangling pointer`?

tasleson replied:
If you enter the function and one or more of the input pointers are null then immediately return without touching any of the out variables.  Otherwise some are undefined and some are changed to unknown.  I see no issue with not changing an out parameter when the inputs are incorrect.  These are pointers which contain the address of an integer, I guess I don't consider them dangling if we don't change their value.

cathay4t replied:
OK. Make sense. Will update.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -400,6 +400,27 @@ typedef enum {
+#define LSM_VOLUME_PHYSICAL_DISK_CACHE_ENABLED      2
+#define LSM_VOLUME_PHYSICAL_DISK_CACHE_DISABLED     3
+#define LSM_VOLUME_PHYSICAL_DISK_CACHE_USE_DISK_SETTING     4
 
```
I'm wondering if we could reduce the number of constants, so that we don't need to have a separate one for each?  Basically have:

```C
#define LSM_VOLUME_CACHE_UNKNOWN  1
#define LSM_VOLUME_CACHE_ENABLED  2
#define LSM_VOLUME_CACHE_DISABLED  3
#define LSM_VOLUME_CACHE_AUTO  4
#define LSM_VOLUME_CACHE_WRITE_BACK  5
#define LSM_VOLUME_CACHE_WRITE_THROUGH  6
#define LSM_VOLUME_CACHE_USE_DISK_SETTING  7
```

What do people think about that?

cathay4t replied:
-1. Then plugin developer or API user have to check document to know which value is available for `write_cache_policy`.

tasleson commented on python_binding/lsm/_client.py:
```diff
@@ -1527,3 +1527,87 @@ def batteries(self, search_key=None, search_value=None, flags=FLAG_RSVD):
+        Version:
+            1.3
+        Usage:
+            Query RAM cache setting and status of certain volume on read
```
certain -> specified

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
 
+/**
+ * New in version 1.3
+ * Change the setting of RAM physical disk cache of certain volume. On some
```
certain -> specified

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+/**
+ * New in version 1.3
+ * Change the setting of RAM physical disk cache of certain volume. On some
+ * product(like HPE SmartArray), this action will be effective at system level
```
products

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+ * New in version 1.3
+ * Change the setting of RAM physical disk cache of certain volume. On some
+ * product(like HPE SmartArray), this action will be effective at system level
+ * which means that even you are requesting a change on a certain volume, this
```
certain -> specific

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+
+/**
+ * New in version 1.3
+ * Change the RAM write cache policy on certain volume. If
```
certain -> specified

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+                                                     lsm_flag flags);
+/**
+ * New in version 1.3
+ * Change the RAM read cache policy of certain volume.
```
certain -> specified

tasleson commented on c_binding/lsm_mgmt.cpp:
```diff
@@ -2716,3 +2716,63 @@ int lsm_volume_cache_info(lsm_connect *c, lsm_volume *volume,
+    if (LSM_FLAG_UNUSED_CHECK(flags)) {
+        return LSM_ERR_INVALID_ARGUMENT;
+    }
+
```
Add check to validate volume

tasleson replied:
validate pdc

tasleson commented on c_binding/lsm_mgmt.cpp:
```diff
@@ -2716,3 +2716,63 @@ int lsm_volume_cache_info(lsm_connect *c, lsm_volume *volume,
+    if (LSM_FLAG_UNUSED_CHECK(flags)) {
+        return LSM_ERR_INVALID_ARGUMENT;
+    }
+
```
Add check to validate volume `LSM_IS_VOL`

tasleson replied:
Validate wcp

tasleson commented on c_binding/lsm_mgmt.cpp:
```diff
@@ -2716,3 +2716,63 @@ int lsm_volume_cache_info(lsm_connect *c, lsm_volume *volume,
+    if (LSM_FLAG_UNUSED_CHECK(flags)) {
+        return LSM_ERR_INVALID_ARGUMENT;
+    }
+
```
Add check to validate volume `LSM_IS_VOL`

tasleson replied:
We should also validate rcp value

tasleson commented:
@cathay4t All the functions in `Change volume RAM cache settings` are using `_set`.  These need to be changed to `_update` per our previous discussion.  `_get` and `_set` are reserved for getting and setting values on an object.

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+ * @param[in] flags         Reserved set to zero
+ * @return LSM_ERR_OK on success else error reason
+ */
+int LSM_DLL_EXPORT lsm_volume_physical_disk_cache_set(lsm_connect *c,
```
_set -> _update

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+ * @param[in] flags         Reserved set to zero
+ * @return LSM_ERR_OK on success else error reason
+ */
+int LSM_DLL_EXPORT lsm_volume_write_cache_policy_set(lsm_connect *c,
```
_set -> _update

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1138,6 +1138,75 @@ int LSM_DLL_EXPORT lsm_volume_cache_info(lsm_connect *c,
+ * @param[in] flags         Reserved set to zero
+ * @return LSM_ERR_OK on success else error reason
+ */
+int LSM_DLL_EXPORT lsm_volume_read_cache_policy_set(lsm_connect *c,
```
_set -> _update

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1122,6 +1122,47 @@ typedef int (*lsm_plug_volume_cache_info) (lsm_plugin_ptr c,
+ * @param[in] flags               Reserved
+ * @return LSM_ERR_OK, else error reason
+ */
+typedef int (*lsm_plug_volume_physical_disk_cache_set) (lsm_plugin_ptr c,
```
_set -> _update

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1122,6 +1122,47 @@ typedef int (*lsm_plug_volume_cache_info) (lsm_plugin_ptr c,
+ * @param[in] flags               Reserved
+ * @return LSM_ERR_OK, else error reason
+ */
+typedef int (*lsm_plug_volume_write_cache_policy_set) (lsm_plugin_ptr c,
```
_set -> _update

tasleson commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1122,6 +1122,47 @@ typedef int (*lsm_plug_volume_cache_info) (lsm_plugin_ptr c,
+ * @param[in] flags               Reserved
+ * @return LSM_ERR_OK, else error reason
+ */
+typedef int (*lsm_plug_volume_read_cache_policy_set) (lsm_plugin_ptr c,
```
_set -> update

tasleson commented on python_binding/lsm/_client.py:
```diff
@@ -1611,3 +1611,164 @@ def volume_cache_info(self, volume, flags=FLAG_RSVD):
         return self._tp.rpc('volume_cache_info', _del_self(locals()))
+
+    @_return_requires(None)
+    def volume_physical_disk_cache_set(self, volume, pdc, flags=FLAG_RSVD):
```
_set -> _update for this and all the other additions

tasleson commented on python_binding/lsm/_client.py:
```diff
@@ -1611,3 +1611,164 @@ def volume_cache_info(self, volume, flags=FLAG_RSVD):
+        Version:
+            1.3
+        Usage:
+            Change the setting of RAM physical disk cache of certain volume. On
```
certain -> specified

tasleson commented on python_binding/lsm/_client.py:
```diff
@@ -1611,3 +1611,164 @@ def volume_cache_info(self, volume, flags=FLAG_RSVD):
+        Version:
+            1.3
+        Usage:
+            Change the RAM write cache policy on certain volume.
```
certain -> specified

tasleson commented on python_binding/lsm/_client.py:
```diff
@@ -1611,3 +1611,164 @@ def volume_cache_info(self, volume, flags=FLAG_RSVD):
+        Version:
+            1.3
+        Usage:
+            Change the RAM read cache policy of certain volume.
```
certain -> specified

tasleson commented on doc/man/lsmcli.1.in:
```diff
@@ -626,6 +626,35 @@ Query RAM cache information for the desired volume.
 \fB--vol\fR \fI<VOL_ID>\fR
 Required. ID of the volume to query cache information.
 
+.SS volume-phy-disk-cache-set
```
-set -> -update

tasleson commented on doc/man/lsmcli.1.in:
```diff
@@ -626,6 +626,35 @@ Query RAM cache information for the desired volume.
 Required. ID of the volume to query cache information.
 
+.SS volume-phy-disk-cache-set
+Disable or enable RAM physical disk cache of certain volume.
```
certain -> specified

tasleson commented on doc/man/lsmcli.1.in:
```diff
@@ -626,6 +626,35 @@ Query RAM cache information for the desired volume.
+Required. \fBEnable\fR or \fBDisable\fR.
+
+.SS volume-read-cache-policy-set
+Disable or enable RAM read cache of certain volume.
```
certain -> specified

tasleson commented on tools/lsmcli/cmdline.py:
```diff
@@ -788,6 +804,31 @@ def _get_item(l, the_id, friendly_name='item', raise_error=True):
         ],
     ),
+    dict(
+        name='volume-phy-disk-cache-set',
```
set -> update

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1044,6 +1045,167 @@ int LSM_DLL_EXPORT lsm_system_read_cache_pct_update(lsm_connect *c,
+ * @param[out]  bs              An array of lsm_battery types
+ * @param[out]  count           Number of batteries
+ * @param[in]   flags           Reserved set to zero
+ * @return LSM_ERR_OK on success else error reason
```
Do we want to follow the new format Blaine established in PR #126 ?

tasleson replied:
Yes, we should.

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt.h:
```diff
@@ -1044,6 +1045,167 @@ int LSM_DLL_EXPORT lsm_system_read_cache_pct_update(lsm_connect *c,
+ * @param[out] physical_disk_cache
+ *                          Whether physical disk's cache is enabled or not.
+ *                          Please be advised, HDD's physical disk ram cache
+ *                          might be not protected by storage system's battery
```
*might not be

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_capabilities.h:
```diff
@@ -115,6 +115,32 @@ typedef enum {
+    LSM_CAP_VOLUME_CACHE_INFO = 57,
+
+    /** Set volume physical disk cache */
+    LSM_CAP_VOLUME_PHYSICAL_DISK_CACHE_SET = 58,
```
Change these capability names to _UPDATE instead of _SET also, keeping in line with the naming change for the operations that can change controller system settings. There are several that would need to change to _UPDATE here.

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1030,6 +1031,138 @@ struct lsm_ops_v1_2 {
+
+/**
+ * New in version 1.3.
+ * Provides for batteries filtering when an storage system doesn't support this
```
Maybe something more like "Allows for battery filtering when a storage system can't support it natively."

cathay4t replied:
I just use the phrase used in `lsm_plug_volume_search_filter`.
I am not English native speaker, should I change all `xxx_search_filter` comments to this?

joehandzik replied:
@cathay4t I think so, yes. Apologies if I missed one, these are long reviews. :)

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1030,6 +1031,138 @@ struct lsm_ops_v1_2 {
+                                                   uint32_t *count);
+/**
+ * New in version 1.3.
+ * Allocate the storage needed for tan array of lsm_battery records.
```
*an

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1030,6 +1031,138 @@ struct lsm_ops_v1_2 {
+
+/**
+ * New in version 1.3.
+ * Change the setting of RAM read disk cache policy of certain volume.
```
@tasleson mentioned this in a couple other comments, but I want to +1 the change to "specific" instead of "certain". There are many instances across these headers that need up be updated that haven't been specifically called out.

joehandzik commented on c_binding/include/libstoragemgmt/libstoragemgmt_plug_interface.h:
```diff
@@ -1039,6 +1172,11 @@ struct lsm_ops_v1_3 {
     lsm_plug_system_read_cache_pct_update sys_read_cache_pct_update;
+    lsm_plug_battery_list battery_list;
+    lsm_plug_volume_cache_info vol_cache_info;
+    lsm_plug_volume_physical_disk_cache_set vol_pdc_set;
```
all these should change to update, also (see line 1174 just above).

joehandzik commented on doc/man/lsmcli.1.in:
```diff
@@ -153,7 +153,7 @@ Required. Valid values are (case insensitive):
 .br
 \fBACCESS_GROUPS\fR, \fBSYSTEMS\fR, \fBDISKS\fR, \fBPLUGINS\fR,
-\fBTARGET_PORTS\fR.
+\fBTARGET_PORTS\fR, \fBBATTEIRES\fR.
```
*BATTERIES

joehandzik commented on tools/lsmcli/cmdline.py:
```diff
@@ -798,6 +859,12 @@ def _get_item(l, the_id, friendly_name='item', raise_error=True):
+    ['ldl', 'local-disk-list'],
+    ['lb', 'list --type batteries'],
+    ['vci', 'volume-cache-info'],
+    ['vcp', 'volume-phy-disk-cache-set'],
```
These last three aliases confuse me. So far, when I create new ones, I've been using the first letter from every work in the operation (see line 860). They're longer, sure, but the user doesn't need to guess which letter actually made it into the alias (here, it's pretty arbitrary that 'volume-phy-disk-cache-set' ended up as vcp. You took the first letters from volume, cache, and phy, but put them in a different order than they exist in the extended command name...as a user, that would frustrate me. I would suggest 'vpdcs' for this specific one, and continue to use the first letter from each word in the extended command for these aliases.

cathay4t replied:
The reason to introduce lsmcli alias is to save key typing[1] and easy to remember.
Personally, `vcp` is easy to type and memorize than `vpdcs` or future `vpdcu`.

Just FYI, using first letter of each word might cause problem for `volume-delete` and `volume-disable`.

[1]: I use lsmcli a lot during coding. So the lsmcli alias do save a lot of my time.
Anyway, I could live with both ways for this trivial issue. @tasleson, any input?

tasleson replied:
I agree with Joe, some consistency here would be good.

cathay4t commented:
V2 patch set uploaded.

Changes in V2:
 * Extra patch to use MEMBER_FUNC_GET.
 * All typo and suggestions mentioned above except:
   * About 'out' goto flag -- I don't want duplication code or dangling pointer.
   * About lsmcli alias -- Just want to keep short alias like `vcp`.

cathay4t commented:
V3 patch set uploaded.

Changes in V3:
 * Changed lsmcli alias to use command's first letter of each world.
 * Removed goto out flag in `lsm_volume_cache_info()` by doing early return on invalid argument.

tasleson commented:
@cathay4t Please re-base and we will get this merged

cathay4t commented:
Rebased.

tasleson commented:
Thanks @cathay4t !

BlaineEXE commented on c_binding/include/libstoragemgmt/libstoragemgmt_types.h:
```diff
@@ -378,6 +383,35 @@ typedef enum {
+    LSM_BATTERY_TYPE_OTHER = 2,
+    LSM_BATTERY_TYPE_CHEMICAL = 3,
+    LSM_BATTERY_TYPE_CAPACITOR = 4,
+} lsm_battery_type;
```
By the convention set elsewhere in the document, shouldn't this be named `lsm_battery_type_type`?

cathay4t replied:
@BlaineEXE Could you point me the document you are referring? From the code I found, `lsm_pool_member_type` and `lsm_disk_type` seems said otherwise.

BlaineEXE replied:
Right above this `lsm_system_mode_type` is used for `LSM_SYSTEM_MODE_...`  when `lsm_system_mode` would do. This seems to indicate you use the `_type` suffix to denote typedef-ed enums. But I also see that `lsm_job_status` is present, which doesn't do this. Looking through the code again, I see that the other `_type` names are appropriate. I think I interpreted incorrectly. Please disregard my original comment.
